### DISCUSSION TITLE
ERROR cannot unmarshal common settings: unexpected end of JSON input

### DISCUSSION THREAD

**Discussion by yefenghu** [qdm12/ddns-updater#251] [Q&A] [answered]
this is openwrt log:

```
stdout: Running version latest built on 2021-09-28T14:25:06Z (commit 963da5a)
stdout: 
stdout: 🔧 Need help? https://github.com/qdm12/ddns-updater/discussions/new
stdout: 🐛 Bug? https://github.com/qdm12/ddns-updater/issues/new
stdout: ✨ New feature? https://github.com/qdm12/ddns-updater/issues/new
stdout: ☕ Discussion? https://github.com/qdm12/ddns-updater/discussions/new
stdout: 💻 Email? quentin.mcgaw@gmail.com
stdout: 💰 Help me? https://www.paypal.me/qmcgaw https://github.com/sponsors/qdm12
stdout: 2021/10/02 03:33:39 ERROR cannot unmarshal common settings: unexpected end of JSON input
stdout: 2021/10/02 03:33:39 INFO Shutdown successful
```

this is my json:

```json
{
    "settings": [
      {
        "provider": "porkbun",
        "domain": "yefenghu.***",
        "host": "@",
        "api_key": "***",
        "secret_api_key": "***",
        "ip_version": "ipv4"
      }
    ]
}
```

**Comment by qdm12** [comment]
Make sure you have bind mounted the file correctly. Usually `unexpected end of JSON input` means it's an empty file or there is a missing bracket somewhere. The JSON you sent looks correct though, maybe try with https://jsonlint.com/ to be sure.

Also maybe the problem comes from Windows line endings `\r\n` in the file? 🤔  (probably not, but maybe)

**Comment by samcro1967** [comment]
@qdm12 I started getting this after the most recent update.  Shows I am running version latest built on 2021-10-15T14:43:47Z (commit 14cd7b0).  My json shows it is valid at https://jsonlint.com.  I get the same error if I change config.json to {} rather than no setting found.  Permissions on config.json are 777 for debugging purposes.  I ran dos2unix on config.json to ensure no Windows characters are in the file.  I also added LOG_LEVEL=debug, but get no additional information.

I am not able to connect to the container to verify it sees the file as the container does not stay running, but my config has not changed. 
 The volume mount in Portainer is as it should be and has always been.  I have been running this for months without any issue.  Any other suggestions on this to try to troubleshoot or help debug?

**logs**
```
========================================
========================================
============= ddns-updater =============
========================================
=========== Made with ❤️ by ============
======= https://github.com/qdm12 =======
========================================
========================================
Running version latest built on 2021-10-15T14:43:47Z (commit 14cd7b0)
🔧 Need help? https://github.com/qdm12/ddns-updater/discussions/new
🐛 Bug? https://github.com/qdm12/ddns-updater/issues/new
✨ New feature? https://github.com/qdm12/ddns-updater/issues/new
☕ Discussion? https://github.com/qdm12/ddns-updater/discussions/new
💻 Email? quentin.mcgaw@gmail.com
💰 Help me? https://www.paypal.me/qmcgaw https://github.com/sponsors/qdm12
2021/10/20 12:48:07 ERROR unexpected end of JSON input
2021/10/20 12:48:07 INFO Shutdown successful
```

**docker-compose**
```
  ddns-updater:
    hostname: ddns-updater
    image: qmcgaw/ddns-updater:latest
    container_name: ddns-updater
    restart: always
    environment:
     - ROOT_URL=/ddns-updater/
     - LOG_LEVEL=debug
     - CONFIG=/updater/data/config.json
    ports:
    - "8103:8000"
    volumes:
     - "${docker}/ddns-updater:/updater/data"
```

**config.json**
```
{
     "settings": [
	    {
			"provider": "duckdns",
			"host": "<redacted>",
			"token": "<redacted>",
			"ip_version": "ipv4",
			"provider_ip": true
		},
		{
			"provider": "cloudflare",
			"zone_identifier": "<redacted>",
			"domain": "<redacted>",
			"host": "@",
			"ttl": 600,
			"token": "<redacted>",
			"ip_version": "ipv4"
		}
	]
}
```

**Reply by qdm12** [reply]
Re-read the documentation on the `CONFIG` variable 😉 Hint: `/updater/data/config.json` is not valid JSON.

However I should add debug logs showing what JSON was read...

**Comment by yefenghu** [comment]
check your config.json is in /root/data/

**Comment by samcro1967** [comment]
@qdm12 Oops.  I added that as another step to troubleshoot today.  Was thinking it was a way to tell the container where to find the file since it does not seem to be finding it.  I have removed it.  That was not there previously and does not work without it.   I am guessing since I do not get no config found with only {} in the config.json file either the container does not see the file or the container is no longer looking in the same spot.

@yefenghu There is no way for me to validate it since the container does not stay running.  The docker-compose says the data dir is /updater/data, not /root/data.  Did it change?  Maybe that is the issue.

**Reply by yefenghu** [reply]
@samcro1967 i do not konw ,i have same log .i solve it by this way

**Comment by samcro1967** [comment]
So if I change my volume to - "${docker}/ddns-updater:/root/data" I do get "WARN Found no setting to update record".  Seems like /updater/data is the right location and it is seeing the file, but is not reading it properly even with just {} in it.  I think permissions are ruled out since they are 777.  Not sure what else it could be at this point.

**Reply by qdm12** [reply]
My volume mount is:

```yml
- ./config.json:/updater/data/config.json:ro
```

Make sure you config.json is not `{}` and that `CONFIG` is not set. It clearly manages to read the config from somewhere, although it looks like it's empty so far.

**Comment by samcro1967** [comment]
It is not empty.  I only did that temporarily to rule out a json issue in my config file.  It is what is above and the ```CONFIG``` environment variable is not set.  That also was temporary.  Mounting the file directly has the same effect.  Without more logging or a way to keep the container running so it can be debugged, not sure there is much more we can do.  It would be good to see the exact path in the debug logs it is trying to mount and any associated errors.

**Volume mount from Portainer**
![image](https://user-images.githubusercontent.com/2319445/138125019-09573206-7e5e-4ead-817e-80eeb739dc3c.png)

**File permissions**
![image](https://user-images.githubusercontent.com/2319445/138125254-dde758ba-5334-4c85-828d-e23b51fd2324.png)

**Docker .env file**
docker=/home/username/Documents/Docker/

**Comment by qdm12** [comment]
Actually, the program does write an empty JSON file `{}` if it doesn't find the file. My guess is it doesn't find the file.

I pushed an image `qmcgaw/ddns-updater:251` with additional INFO and DEBUG logs regarding where the config is read from, what is read and if a new config file is created. Please try with this one?

**Comment by samcro1967** [comment]
Log looks the same other than no version listed.  I have env var LOG_LEVEL=debug.

```
========================================
========================================
============= ddns-updater =============
========================================
=========== Made with ❤️ by ============
======= https://github.com/qdm12 =======
========================================
========================================

Running version unknown built on an unknown date (commit unknown)

🔧 Need help? https://github.com/qdm12/ddns-updater/discussions/new
🐛 Bug? https://github.com/qdm12/ddns-updater/issues/new
✨ New feature? https://github.com/qdm12/ddns-updater/issues/new
☕ Discussion? https://github.com/qdm12/ddns-updater/discussions/new
💻 Email? quentin.mcgaw@gmail.com
💰 Help me? https://www.paypal.me/qmcgaw https://github.com/sponsors/qdm12
2021/10/20 22:47:32 ERROR unexpected end of JSON input
2021/10/20 22:47:32 INFO Shutdown successful
```

**Reply by qdm12** [reply]
Try with `LOG_CALLER=short`

**Comment by samcro1967** [comment]
That did not change the logging level either.

**Reply by qdm12** [reply]
Well that's not normal. Are you sure you set `LOG_CALLER` and not just `LOG_LEVEL` as I mentioned?

Running `docker run -it --rm -e LOG_CALLER=short qmcgaw/ddns-updater:251` gives me

```
========================================
========================================
============= ddns-updater =============
========================================
=========== Made with ❤️ by ============
======= https://github.com/qdm12 =======
========================================
========================================

Running version unknown built on an unknown date (commit unknown)

🔧 Need help? https://github.com/qdm12/ddns-updater/discussions/new
🐛 Bug? https://github.com/qdm12/ddns-updater/issues/new
✨ New feature? https://github.com/qdm12/ddns-updater/issues/new
☕ Discussion? https://github.com/qdm12/ddns-updater/discussions/new
💻 Email? quentin.mcgaw@gmail.com
💰 Help me? https://www.paypal.me/qmcgaw https://github.com/sponsors/qdm12
2021/10/20 23:50:03 json.go:46: INFO reading JSON config from file /updater/data/config.json
2021/10/20 23:50:03 json.go:53: INFO file not found, creating an empty settings file
2021/10/20 23:50:03 main.go:187: WARN Found no setting to update record
2021/10/20 23:50:03 main.go:273: INFO backup: disabled
2021/10/20 23:50:03 server.go:48: INFO http server: listening on :8000
2021/10/20 23:50:03 server.go:45: INFO healthcheck server: listening on 127.0.0.1:9999
```

**Comment by samcro1967** [comment]
Here is the docker-compose I used.

```
  ddns-updater:
    hostname: ddns-updater
    image: qmcgaw/ddns-updater:251
    container_name: ddns-updater
    restart: always
    environment:
     - ROOT_URL=/ddns-updater/
     - LOG_LEVEL=debug
     - LOG_CALLER=short
    ports:
    - "8103:8000"
    volumes:
     - "${docker}/ddns-updater:/updater/data/"
```

**Reply by qdm12** [reply]
If you don't see the Source go file and the line as I do, then this config is probably not used in your container. I'm not sure how I can help at this point...

**Comment by samcro1967** [comment, marked as answer]
I can't explain it, but this morning I removed the container and both images, rebooted, and reran docker-compose up and it is working.  I had tried removing the container and image yesterday, so It was either the reboot or the combination of the two.  I was obviously running the new image.  Seems like something between docker or that image and the file system, some type of cache maybe.  I've got 50+ other containers running without issue.  Very strange.  Appreciate your help.

**Reply by qdm12** [reply]
Yes I encountered similar stories with users of Portainer on various of my images. It seems sometimes it doesn't update the settings anymore and removing the container is the solution. I'm happy it works for you now!

**Comment by yefenghu** [comment]
this is my log:

```
_____________________________________________________________________________________________________________
stdout: 2021/10/29 16:07:01 INFO reading JSON config from file /updater/data/config.json
stdout: 2021/10/29 16:07:01 INFO Found single setting to update record
stdout: 2021/10/29 16:07:05 INFO Reading history from database: domain yefenghu.* host @
stdout: 2021/10/29 16:07:05 INFO backup: disabled
stdout: 2021/10/29 16:07:05 INFO healthcheck server: listening on 127.0.0.1:9999
stdout: 2021/10/29 16:07:05 INFO http server: listening on :8000
stdout: 2021/10/29 16:12:05 WARN obtaining ipv4 address: try 1 of 3: no TXT record found
stdout: 2021/10/29 16:12:06 INFO obtaining ipv4 address: succeeded after 2 tries
stdout: 2021/10/29 16:17:05 INFO IPv4 address of yefenghu.* is 175.11.*.* and your IPv4 address is 240e:5a:*:0:114:114:114:102
stdout: 2021/10/29 16:17:05 INFO Updating record [domain: yefenghu.* | host: @ | provider: Porkbun] to use 240e:5a:6805:0:114:114:114:102
stdout: 2021/10/29 16:27:05 WARN obtaining ipv4 address: try 1 of 3: no TXT record found
stdout: 2021/10/29 16:27:05 INFO obtaining ipv4 address: succeeded after 2 tries
stdout: 2021/10/29 16:32:05 INFO IPv4 address of yefenghu.* is 175.11.*.* and your IPv4 address is 240e:5a*:0:114:114:114:102
stdout: 2021/10/29 16:32:05 INFO Updating record [domain: yefenghu.* | host: @ | provider: Porkbun] to use 240e:5a:6805:0:114:114:114:102
stdout: 2021/10/29 16:32:07 ERROR bad HTTP status: 400: {"status":"ERROR","message":"Edit error: We were unable to edit the DNS record."}
__________________________________________________________
```

175.11.*.* is correct , 240e:5a*:0:114:114:114:102 i do not konw what it from,need to solve it ?

**Reply by yefenghu** [reply]
this is later log
——————————————————————————————————————————————————
stdout: 2021/10/29 19:32:05 INFO IPv4 address of yefenghu.cyou is 175.11.*.* and your IPv4 address is 58.217.*.*
stdout: 2021/10/29 19:32:05 INFO Updating record [domain: yefenghu.* | host: @ | provider: Porkbun] to use 58.217.*.*
stdout: 2021/10/29 19:32:07 ERROR bad HTTP status: 400: {"status":"ERROR","message":"Edit error: We were unable to edit the DNS record."}
stdout: 2021/10/29 19:42:05 WARN obtaining ipv4 address: try 1 of 3: no TXT record found
________________________________________________________________________________________________________________________
175.11.. is correct,240e:5a*:0:114:114:114:102 change into 58.217.*.*,i do not anything .

**Reply by qdm12** [reply]
That's an IPv6 address. Please create an issue together with your config (without credentials) I'll try to address it this weekend. The program shouldn't find an IPv6 address when querying an ipv4 address. Maybe try setting ip_version to `ipv4` in your config for now, that might solve it until I fix it for the default ipv4_or_v6.

**Reply by qdm12** [reply]
I don't think `yefenghu.*` is valid domain for Porkbun 🤔
